Add ecdh_curves support

kfaseela opened the pull request:
Fixes https://github.com/istio/istio/issues/41645

Signed-off-by: Faseela K <faseela.k@est.tech>

istio-testing commented:
@kfaseela: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
release-notes_api | 4ff34ac01bd84144cbaba8dae32b08a4cdaa9f34 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_api/2543/release-notes_api/1588253112429187072) | false | `/test release-notes`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@kfaseela: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
gencheck_api | eec2bc317d20972f340a9f6b6e1f9a0087331571 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_api/2543/gencheck_api/1591092776496271360) | true | `/test gencheck`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn commented:
/retest

istio-testing commented:
@kfaseela: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
gencheck_api | eec2bc317d20972f340a9f6b6e1f9a0087331571 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_api/2543/gencheck_api/1591100638668787712) | true | `/test gencheck`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

costinm commented on networking/v1alpha3/destination_rule.proto:
```diff
@@ -1125,6 +1125,11 @@ message ClientTLSSettings {
   google.protobuf.BoolValue insecure_skip_verify = 8;
+  //
+  // Optional: If specified, the TLS connection will only support the specified ECDH curves.
+  // If not specified, the default curves will be used. For additional details, refer to
```
Use case for having this as a per-host option ? What does it even mean - reject the cert if it uses the curves ? For the leaf - or the entire chain ? 

I really, really don't think this kind of low level config belongs in DR. Maybe as a global option in MeshConfig or env variable - far too narrow use case and far too confusing. 

Even there - I don't think going to the level of ECDH curves is right - I think we have a list of ciphers, we can add in the same place a list of cert keys to support, and maybe a separate one for root keys.

kfaseela replied:
Hello Costin, as per 3GPP standard, [3GPP TS 33.210](https://portal.3gpp.org/desktopmodules/Specifications/SpecificationDetails.aspx?specificationId=2279), section 6.2 - we need a mechanism to allow/disallow certain ecdh_curves for external traffic. While currently it may be fine to have this in the MeshConfig, we already have ciphersuites in ServerTLSSettings, and an attempt to add cipher to DR is also in progress. So thought ecdh_curves can also go in both the places. Also there can be a possibility of multiple applications in future requiring separate ecdh_curves and hence thought it is better to be future proof. Envoy already supports this config.

louiscryan replied:
Can you provide more context on the linked specification.

Envoy only supports two curves today P256 and X22591 (?) and if FIPS is enabled then only P256

Do you want to support more options than this ? If so what might they be ?

FIPS may support more over time but TLS 1.3 is also trying to restrict the choices IIRC so there's an interesting balancing act here

See
https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Ar3.pdf. [Appendix D]
https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml#tls-parameters-8. (see recommended col)

kfaseela replied:
@louiscryan : 3GPP is the 5G telco standard (https://www.3gpp.org/). We work with this as the base for all Telco requirements, and as I mentioned before 3GPP TS 33.210 section 6.2 states that 
"
For TLS 1.3 ==> 
        The requirements given in section 9.1 of TLS 1.3 RFC 8446 [66] shall be followed. In addition:
 	Key exchange with secp384r1 should be supported.

For TLS1.2 ==>
-	For ECDHE, the curve secp256r1 (P-256) as defined in RFC 8422 [71] shall be supported, secp384r1 (P-384) as defined in RFC 8422 [71] should be supported.  Except curve25519, ed25519, and W-25519, elliptic curve groups of less than 256 bits shall not be supported.
-	For DHE, Diffie-Hellman groups of at least 4096 bits should be supported. Diffie-Hellman groups smaller than 2048 bits shall not be supported.
"
Since there is a mention of something of what is allowed and what is not allowed, isn't it better to have a possibility at istio to make this configurable? Ofcourse if you think that it is too low-level for DR/Gateway, may be a MeshConfig option for controlling the same for external traffic would be good.

And for the values supported in envoy, I think the values you have mentioned are default values, and the entire list is seen here? 
https://github.com/envoyproxy/envoy/blob/main/source/extensions/transport_sockets/tls/context_impl.cc#L328 (correct me if I am wrong).

jewertow replied:
> Use case for having this as a per-host option ?

Some our customers have to connect to external services that require specific ciphers and ECDH curves. On the other hand, others have to use specific TLS parameters in the entire mesh, because of their company security policies.

costinm commented on networking/v1alpha3/gateway.proto:
```diff
@@ -732,4 +732,10 @@ message ServerTLSSettings {
   // Otherwise default to the default cipher list supported by Envoy.
   repeated string cipher_suites = 9;
+
+  //
```
Same comment as above. 

If this is really about FIPS compliance - having a MeshConfig "FIPS" bool would be a much better UX. 

If it is not about FIPS - I would love to see a use case for every user creating a Gateway to know the names of the 3 curves supported by TLS and manually putting them in each and every yaml file and DR ( and getting locked in Istio GW API - since the new K8S Gateway lacks this feature in the core API )

costinm commented:
I don't think either DestinationRule nor Gateway should have this kind of
low-level, arcane config.
Most developers who work with DR and Gateway are unlikely to know if FIPS
is required or what is the meaning of each curve.
I don't think adding ciphers was right either.

A mesh admin may want to set a 'FIPS setting' or some similar high level
configs, I am not against allowing
TLS config - but not so low level, in particular for something that only
has 2 options.

I don't think we even support generating ED22591 certs - we could just
support EC256, and users who want ED
can use EnvoyFilter.





On Mon, Nov 14, 2022 at 4:51 PM Louis Ryan ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In networking/v1alpha3/destination_rule.proto
> <https://github.com/istio/api/pull/2543#discussion_r1022207257>:
>
> > @@ -1125,6 +1125,11 @@ message ClientTLSSettings {
>    // be `true` by default in a later version where, going forward, it will be
>    // enabled by default.
>    google.protobuf.BoolValue insecure_skip_verify = 8;
> +  //
> +  // Optional: If specified, the TLS connection will only support the specified ECDH curves.
> +  // If not specified, the default curves will be used. For additional details, refer to
>
> Can you provide more context on the linked specification.
>
> Envoy only supports two curves today P256 and X22591 (?) and if FIPS is
> enabled then only P256
>
> Do you want to support more options than this ? If so what might they be ?
>
> FIPS may support more over time but TLS 1.3 is also trying to restrict the
> choices IIRC so there's an interesting balancing act here
>
> See
> https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Ar3.pdf.
> [Appendix D]
>
> https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml#tls-parameters-8.
> (see recommended col)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/api/pull/2543#discussion_r1022207257>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAUR2W3OBOCHVG75JYN4FTWILNCFANCNFSM6AAAAAARWPL53Q>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

kfaseela commented:
> I don't think either DestinationRule nor Gateway should have this kind of low-level, arcane config. Most developers who work with DR and Gateway are unlikely to know if FIPS is required or what is the meaning of each curve. I don't think adding ciphers was right either. A mesh admin may want to set a 'FIPS setting' or some similar high level configs, I am not against allowing TLS config - but not so low level, in particular for something that only has 2 options. I don't think we even support generating ED22591 certs - we could just support EC256, and users who want ED can use EnvoyFilter.
> […](#)
> On Mon, Nov 14, 2022 at 4:51 PM Louis Ryan ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In networking/v1alpha3/destination_rule.proto <[#2543 (comment)](https://github.com/istio/api/pull/2543#discussion_r1022207257)>: > @@ -1125,6 +1125,11 @@ message ClientTLSSettings { // be `true` by default in a later version where, going forward, it will be // enabled by default. google.protobuf.BoolValue insecure_skip_verify = 8; + // + // Optional: If specified, the TLS connection will only support the specified ECDH curves. + // If not specified, the default curves will be used. For additional details, refer to Can you provide more context on the linked specification. Envoy only supports two curves today P256 and X22591 (?) and if FIPS is enabled then only P256 Do you want to support more options than this ? If so what might they be ? FIPS may support more over time but TLS 1.3 is also trying to restrict the choices IIRC so there's an interesting balancing act here See https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Ar3.pdf. [Appendix D] https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml#tls-parameters-8. (see recommended col) — Reply to this email directly, view it on GitHub <[#2543 (comment)](https://github.com/istio/api/pull/2543#discussion_r1022207257)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAAUR2W3OBOCHVG75JYN4FTWILNCFANCNFSM6AAAAAARWPL53Q> . You are receiving this because you commented.Message ID: ***@***.***>

@costinm @louiscryan: The envoy code below, shows a list of ecdh_curves and not two. That is why we needed a mechanism to limit. The two values you mention, I am assuming they are the default values? (I may be wrong, please correct if so. I understand that big API changes may not be intended for a value which has just 2 options). 
https://github.com/envoyproxy/envoy/blob/main/source/extensions/transport_sockets/tls/context_impl.cc#L328

nrjpoddar commented:
So we support specifying Cipher list only for GWs and for Mesh just specifying the minimum TLS version.

For augmenting DR we have had lengthy debates in the past so if you can clearly mention your use case where you want this capability per host or per workload or mesh wide it would be easier to figure out where we should add this option.

kfaseela commented:
> So we support specifying Cipher list only for GWs and for Mesh just specifying the minimum TLS version.
> 
> For augmenting DR we have had lengthy debates in the past so if you can clearly mention your use case where you want this capability per host or per workload or mesh wide it would be easier to figure out where we should add this option.

@nrjpoddar : Per host level would be more future proof, but as per the current requirement it is needed mesh-wide only. So would it be acceptable if I modify the PR to have the config in MeshConfig?

costinm commented:
No concern with having this in MeshConfig. It can probably live in
ProxyConfig too - and we can apply the per-workload or namespace
default/override
mechanisms. Just not DR. For gateway I am less concerned since users will
likely stop using it as they migrate to the official K8S Gateway.

I would still appreciate ( and all Istio users who don't know what 25519 or
256 curve is ) to also add a "FIPS" bool, and if the
FIPS bool is set just change the curves and ciphers to whatever is required
(unless user explicitly overrides ).
Not sure what is the current FIPS requirement for certs - but we may also
use that to default to EC256 certs (supported for some time)
instead of RSA ?


On Wed, Nov 16, 2022 at 6:26 AM Faseela K ***@***.***> wrote:

> So we support specifying Cipher list only for GWs and for Mesh just
> specifying the minimum TLS version.
>
> For augmenting DR we have had lengthy debates in the past so if you can
> clearly mention your use case where you want this capability per host or
> per workload or mesh wide it would be easier to figure out where we should
> add this option.
>
> @nrjpoddar <https://github.com/nrjpoddar> : Per host level would be more
> future proof, but as per the current requirement it is needed mesh-wide
> only. So would it be acceptable if I modify the PR to have the config in
> MeshConfig?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/api/pull/2543#issuecomment-1317106515>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAUR2VNOQ5DGFQIGUK44XTWITVIZANCNFSM6AAAAAARWPL53Q>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

howardjohn commented:
AFAIK Envoy FIPS requires a completely different compiled binary so its hard to have just a plain "FIPS" flag to use it

ramaraochavali commented:
yes it needs a completely new binary. I think enabling "FIPS" flag might be confusing to users

jewertow commented:
> And we end up configuring EnvoyFilters for all ingress/egress external traffic to comply to the 3GPP spec, which gets really complicated, especially when we use sidecar ingress/egress TLS as well.

I totally agree with @kfaseela. EnvoyFilters are very difficult to maintain and support.

jewertow commented:
> It can probably live in ProxyConfig.

It also sounds good to me. Anything but EnvoyFilter.

costinm commented:
Are we building istio-proxy images with the FIPS binary ? If yes - the flag could also switch the injection to use that - so it's not confusing and works with minimal config ?

My goal here is to not expose the users to arcane crypto configs that few people understand. We can call it something else - 'safeCrypto', etc. But in practice almost everyone messing with the settings really wants FIPS, so if we can get the binary and all the obscure settings in one bool - it would likely be best UX for a lot of people.

jewertow commented:
> My goal here is to not expose the users to arcane crypto configs that few people understand.

But often those few people decide whether to deploy Istio in a particular system or not. You're right that only a few people understand these settings and developers don't decide which TLS settings they choose, but architects/admins still need an ability to apply specific TLS configurations.

jewertow commented:
I think that ProxyConfig is a way to go, because it will allow us to apply specific TLS settings mesh-wide, namespaces-wide or to a particular workload, so it would cover all cases. What do you think?

ramaraochavali commented:
> Are we building istio-proxy images with the FIPS binary ? If yes - the flag could also switch the injection to use that - so it's not confusing and works with minimal config ?

No. We are not.

jewertow commented:
I forgot that these settings may still be needed for mTLS inside a mesh, so then DestinationRule would have to implement these settings as well. Taking this into account, MeshConfig and Gateway are better place than ProxyConfig.

howardjohn commented:
Don't we already have MeshConfig_TLSConfig_TLSProtocol for within the mesh

On Fri, Nov 18, 2022 at 3:46 AM Jacek Ewertowski ***@***.***>
wrote:

> I forgot that these settings may still be needed for mTLS inside a mesh,
> so then DestinationRule would have to implement these settings as well.
> Taking this into account, MeshConfig and Gateway are better place than
> ProxyConfig.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/api/pull/2543#issuecomment-1319893804>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEYGXM57BWAGERERZAF2FLWI5UBXANCNFSM6AAAAAARWPL53Q>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

jewertow commented:
Yes, we have.

costinm commented:
On Fri, Nov 18, 2022 at 3:27 AM Jacek Ewertowski ***@***.***>
wrote:

> My goal here is to not expose the users to arcane crypto configs that few
> people understand.
>
> But often those few people decide whether to deploy Istio in a particular
> system or not. You're right that only a few people understand these
> settings and developers don't decide which TLS settings they choose, but
> architects/admins still need an ability to apply specific TLS
> configurations.
>

Yes, but we can't expect the architects/admins that know the difference
between EC curves to edit or review every single DestinationRule where a
developer may
set  curves.

I am not saying the crypto experts should not be able to fine tune TLS -
quite the opposite, it is very important. Just not in APIs that mere
mortals use.

We have TLS settings in DR to cover common cases where a developer may have
to do configs - self signed or 'enterprise CAs', external
sites where hostname and cert don't match, etc. Even this resulted in lots
of problems and confusion. That doesn't mean everything
that can be configured in TLS should be in DR.

We messed up early on by creating dozens of unnecessary CRDs for Mixer -
and we moved to the other extreme of never separating
configs. Which is accidentally ok - it forces us to work with other vendors
in the K8S space and to a much higher bar for API review.
But in the end I hope all those fine TLS settings will end up in a
TLS-specific CRD designed for TLS experts and shared by
all K8S components using TLS. Meanwhile - MeshConfig is good enough.

Message ID: ***@***.***>
>

kfaseela commented:
> On Fri, Nov 18, 2022 at 3:27 AM Jacek Ewertowski ***@***.***> wrote: My goal here is to not expose the users to arcane crypto configs that few people understand. But often those few people decide whether to deploy Istio in a particular system or not. You're right that only a few people understand these settings and developers don't decide which TLS settings they choose, but architects/admins still need an ability to apply specific TLS configurations.
> Yes, but we can't expect the architects/admins that know the difference between EC curves to edit or review every single DestinationRule where a developer may set curves. I am not saying the crypto experts should not be able to fine tune TLS - quite the opposite, it is very important. Just not in APIs that mere mortals use. We have TLS settings in DR to cover common cases where a developer may have to do configs - self signed or 'enterprise CAs', external sites where hostname and cert don't match, etc. Even this resulted in lots of problems and confusion. That doesn't mean everything that can be configured in TLS should be in DR. We messed up early on by creating dozens of unnecessary CRDs for Mixer - and we moved to the other extreme of never separating configs. Which is accidentally ok - it forces us to work with other vendors in the K8S space and to a much higher bar for API review. But in the end I hope all those fine TLS settings will end up in a TLS-specific CRD designed for TLS experts and shared by all K8S components using TLS. Meanwhile - MeshConfig is good enough. Message ID: ***@***.***>
> […](#)

@costinm @jewertow @ramaraochavali : I have removed ecdh_curves from DR, and added to MeshConfig. I have retained in the ServerTLSSettings on Gateway if that would be okay. Let me know your comments

jewertow commented on mesh/v1alpha1/config.proto:
```diff
@@ -1119,13 +1119,33 @@ message MeshConfig {
+  //   egress:
+  //     ecdhCurves:
+  //       - P-256
+  //       - P-512
```
This is what I was suggesting, but now I see that it only solves problem for mesh mTLS and ingress. In case of egress traffic it won't solve the problem, because:
- in `PASSTHROUGH` mode `ecdhCurves` makes no effect - there is no TLS negotiation, traffic leaves mesh as is;
- in `ISTIO_MUTUAL` mode `ecdhCurves` would be effective only between a client and a gateway (assuming that `meshMTLS` supports ecdh curves), but traffic leaving the gateway will not use these curves;
- in `SIMPLE`/`MUTUAL` mode - same as in case of ISTIO_MUTUAL, except that TLS settings would have to be configured by an application, not by a sidecar.

So taking above into account, I think that the only way to achieve the goal of using specific ciphers or ECDH curves for external traffic, we have to use DestinationRule.

jewertow replied:
For me the most important is how we will document these fields. I would like to clarify that egress settings are relevant for TLS client or UpstreamTlsContext and are used to encrypt traffic leaving the mesh. It's important, because in case of ingress it's obvious and intuitive that settings are applicable for downstream traffic. In case of egress, we have to explicitly say that it's for traffic leaving the mesh after TLS termination. Otherwise someone may consider this field as a setting for DownstreamTlsContext in egress gateway.

But please don't update your code until others share opinions.

costinm replied:
Same as above - does this mean only client certs with those curves will be accepted ? Or is it specifically for the DH exchange ? 

Should we express this in TLS terms - the list of algorithms includes both cipher and exchange, and may be more familiar to people configuring TLS.

kyessenov replied:
@costinm These should manifest as DH parameters KeyExchanges message, for the ephemeral elliptic curve DH exchange. Curves are separate from cipher suites, which are in the Hello messages I think.

jewertow commented:
> I have retained in the ServerTLSSettings on Gateway if that would be okay.

I just want to note that it will be effective only for ingress traffic.

howardjohn commented on mesh/v1alpha1/config.proto:
```diff
@@ -1119,13 +1119,34 @@ message MeshConfig {
+  // Please see [ServerTLSSettings] (https://istio.io/latest/docs/reference/config/networking/gateway/#ServerTLSSettings)
+  // to override the global configuration specified here.
+  TLSConfig external_ingress = 64;
+  // Egress gateways do not respect min TLS version configured here currently.
```
There is no such thing as an "egress" and "ingress" Gateway in Istio APIs. These are merely use cases and names we recommend users. So this cannot really be implemented

howardjohn replied:
Also what about egress from sidecars?

kfaseela replied:
@howardjohn  I made this as a single `mesh_external_TLS` config. Also updated the field comments. Let me know if it is better now.

jewertow replied:
> There is no such thing as an "egress" and "ingress" Gateway in Istio APIs.

You're right, but these names are used in Sidecar API. I'm also not very convinced to these names, just explain where the idea came from.

> So this cannot really be implemented.

I assumed that ingress and egress settings would be applied to inbound listeners and clusters respectively.

> Also what about egress from sidecars?

If settings were applied to clusters, it wouldn't matter if traffic leaves mesh from a sidecar or a gateway. It would work in both cases.

howardjohn replied:
My initial concerns is because this all about "egress gateway" and "ingress gateway" according to the comments. What is documented in the proto currently is not possible to implemented.

jewertow replied:
Oh yeah, right... I totally forgot that we are not able to determine if a given listener will be applied to an ingress or egress gateway.

jewertow replied:
Thank you for pointing this out.

costinm commented:
Does it really matter ? I think we are again falling into the trap of
adding APIs because we can, not because we have to.

If you have a real customer or a real use case where EC256 would not be
allowed for ingress, only for egress - I would really like to understand
why.
I know some people don't trust EC256, while others require it. But it's
either required by some rule - or forbidden, never heard of
'ok to use it but only for some things'.



On Mon, Nov 28, 2022 at 11:44 AM Jacek Ewertowski ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In mesh/v1alpha1/config.proto
> <https://github.com/istio/api/pull/2543#discussion_r1033981289>:
>
> >    TLSConfig mesh_mTLS = 63;
> +  // Ingress gateways do not respect min TLS version configured here currently.
> +  // Please see [ServerTLSSettings] (https://istio.io/latest/docs/reference/config/networking/gateway/#ServerTLSSettings)
> +  // to override the global configuration specified here.
> +  TLSConfig external_ingress = 64;
> +  // Egress gateways do not respect min TLS version configured here currently.
>
> Thank you for pointing this out.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/api/pull/2543#discussion_r1033981289>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAUR2XV4ITMK7DVPS5FAD3WKUDTPANCNFSM6AAAAAARWPL53Q>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

kfaseela commented:
> Does it really matter ? I think we are again falling into the trap of adding APIs because we can, not because we have to. If you have a real customer or a real use case where EC256 would not be allowed for ingress, only for egress - I would really like to understand why. I know some people don't trust EC256, while others require it. But it's either required by some rule - or forbidden, never heard of 'ok to use it but only for some things'.
> […](#)
> On Mon, Nov 28, 2022 at 11:44 AM Jacek Ewertowski ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In mesh/v1alpha1/config.proto <[#2543 (comment)](https://github.com/istio/api/pull/2543#discussion_r1033981289)>: > TLSConfig mesh_mTLS = 63; + // Ingress gateways do not respect min TLS version configured here currently. + // Please see [ServerTLSSettings] (https://istio.io/latest/docs/reference/config/networking/gateway/#ServerTLSSettings) + // to override the global configuration specified here. + TLSConfig external_ingress = 64; + // Egress gateways do not respect min TLS version configured here currently. Thank you for pointing this out. — Reply to this email directly, view it on GitHub <[#2543 (comment)](https://github.com/istio/api/pull/2543#discussion_r1033981289)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAAUR2XV4ITMK7DVPS5FAD3WKUDTPANCNFSM6AAAAAARWPL53Q> . You are receiving this because you were mentioned.Message ID: ***@***.***>

After repeated internal discussions, I have finally got a confirmation, and agree to you @costinm that there is no current usecase for separate values for ingress and egress. So a single configurable option for ecdh curves should be good enough. Does that sound okay to you? If so I will push a revised PR making the changes.

costinm commented:
Thanks, and if use cases are found - we can open this again ( and I believe
EnvoyFilter provides an escape until this happens ).



On Wed, Jan 11, 2023 at 3:24 AM Faseela K ***@***.***> wrote:

> Does it really matter ? I think we are again falling into the trap of
> adding APIs because we can, not because we have to. If you have a real
> customer or a real use case where EC256 would not be allowed for ingress,
> only for egress - I would really like to understand why. I know some people
> don't trust EC256, while others require it. But it's either required by
> some rule - or forbidden, never heard of 'ok to use it but only for some
> things'.
> … <#m_2639346283528683268_>
> On Mon, Nov 28, 2022 at 11:44 AM Jacek Ewertowski *@*.*> wrote: @.**
> commented on this pull request. ------------------------------ In
> mesh/v1alpha1/config.proto <#2543 (comment)
> <https://github.com/istio/api/pull/2543#discussion_r1033981289>>: >
> TLSConfig mesh_mTLS = 63; + // Ingress gateways do not respect min TLS
> version configured here currently. + // Please see [ServerTLSSettings] (
> https://istio.io/latest/docs/reference/config/networking/gateway/#ServerTLSSettings)
> + // to override the global configuration specified here. + TLSConfig
> external_ingress = 64; + // Egress gateways do not respect min TLS version
> configured here currently. Thank you for pointing this out. — Reply to this
> email directly, view it on GitHub <#2543 (comment)
> <https://github.com/istio/api/pull/2543#discussion_r1033981289>>, or
> unsubscribe
> https://github.com/notifications/unsubscribe-auth/AAAUR2XV4ITMK7DVPS5FAD3WKUDTPANCNFSM6AAAAAARWPL53Q
> . You are receiving this because you were mentioned.Message ID: *@*.***>
>
> After repeated internal discussions, I have finally got a confirmation,
> and agree to you @costinm <https://github.com/costinm> that there is no
> current usecase for separate values for ingress and egress. So a single
> configurable option for ecdh curves should be good enough. Does that sound
> okay to you? If so I will push a revised PR making the changes.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/api/pull/2543#issuecomment-1378604378>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAUR2X6DQGZIDOLO7G4YZLWR2J7BANCNFSM6AAAAAARWPL53Q>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

kfaseela commented:
> Thanks, and if use cases are found - we can open this again ( and I believe EnvoyFilter provides an escape until this happens ).
> […](#)
> On Wed, Jan 11, 2023 at 3:24 AM Faseela K ***@***.***> wrote: Does it really matter ? I think we are again falling into the trap of adding APIs because we can, not because we have to. If you have a real customer or a real use case where EC256 would not be allowed for ingress, only for egress - I would really like to understand why. I know some people don't trust EC256, while others require it. But it's either required by some rule - or forbidden, never heard of 'ok to use it but only for some things'. … <#m_2639346283528683268_> On Mon, Nov 28, 2022 at 11:44 AM Jacek Ewertowski *@*.*> wrote: @.** commented on this pull request. ------------------------------ In mesh/v1alpha1/config.proto <#2543 (comment) <[#2543 (comment)](https://github.com/istio/api/pull/2543#discussion_r1033981289)>>: > TLSConfig mesh_mTLS = 63; + // Ingress gateways do not respect min TLS version configured here currently. + // Please see [ServerTLSSettings] ( https://istio.io/latest/docs/reference/config/networking/gateway/#ServerTLSSettings) + // to override the global configuration specified here. + TLSConfig external_ingress = 64; + // Egress gateways do not respect min TLS version configured here currently. Thank you for pointing this out. — Reply to this email directly, view it on GitHub <#2543 (comment) <[#2543 (comment)](https://github.com/istio/api/pull/2543#discussion_r1033981289)>>, or unsubscribe https://github.com/notifications/unsubscribe-auth/AAAUR2XV4ITMK7DVPS5FAD3WKUDTPANCNFSM6AAAAAARWPL53Q . You are receiving this because you were mentioned.Message ID: *@*.***> After repeated internal discussions, I have finally got a confirmation, and agree to you @costinm <https://github.com/costinm> that there is no current usecase for separate values for ingress and egress. So a single configurable option for ecdh curves should be good enough. Does that sound okay to you? If so I will push a revised PR making the changes. — Reply to this email directly, view it on GitHub <[#2543 (comment)](https://github.com/istio/api/pull/2543#issuecomment-1378604378)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAAUR2X6DQGZIDOLO7G4YZLWR2J7BANCNFSM6AAAAAARWPL53Q> . You are receiving this because you were mentioned.Message ID: ***@***.***>

@costinm I meant I will retain the changes in this PR for MeshConfig option, but without the need for separate values for ingress and egress. Also no need for DR side changes etc. Is that okay? :)

istio-testing commented:
@kfaseela: PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

istio-testing commented:
@kfaseela: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
build_api | 41604e8158292e067918505d3c289c9e75d88902 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_api/2543/build_api/1617906255748141056) | true | `/test build`
gencheck_api | 41604e8158292e067918505d3c289c9e75d88902 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_api/2543/gencheck_api/1617906255785889792) | true | `/test gencheck`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@kfaseela: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
build_api | 41604e8158292e067918505d3c289c9e75d88902 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_api/2543/build_api/1617906255748141056) | true | `/test build`
gencheck_api | e73e7400d6ff71e52aab83dfc4e9d073939fb030 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_api/2543/gencheck_api/1617907007296114688) | true | `/test gencheck`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

ramaraochavali commented:
This LGTM. Will wait for API PR to be merged and I Will take a final look and approve

howardjohn commented on mesh/v1alpha1/config.proto:
```diff
@@ -1117,13 +1117,34 @@ message MeshConfig {
+  // Mesh external does not respect min TLS version configured here currently.
+  // This global configuration can be overridden per Gateway by configuring
+  // [ServerTLSSettings](https://istio.io/latest/docs/reference/config/networking/gateway/#ServerTLSSettings)
+  TLSConfig mesh_external_TLS = 64;
```
What exactly is "mesh_external_TLS"? Is this only for terminating traffic (Gateway)? So its a global setting for Gateway TLS?

Does it also apply to non-mTLS Destinationrule origination?

kfaseela replied:
It means traffic incoming/leaving the mesh. Because mesh_mtls is already there in MeshConfig at line 1138 and it has config parameters applicable within the mesh, so did not touch that.

howardjohn replied:
So its really "Any TLS *except* ISTIO_MUTUAL mTLS" then? Can we make this clear in the API docs

kfaseela replied:
I have updated the docs

ramaraochavali replied:
> So its really "Any TLS except ISTIO_MUTUAL mTLS" then? Can we make this clear in the API docs

I think this is very confusing.  In side mesh, we use MUTUAL TLS. To keep that case under MeshExternal traffic is not correct.

If our intent is to just enable this for any TLS that is not ISTIO_MUTUAL, why not just keep it under `mesh_mTLS` and document that it will only be applied for non ISTIO MUTUAL TLS.  That way it is easy to understand. In the current state -  it is difficult to reason about and implement. How do we know which traffic is egress from sidecar and how do we know which traffic coming in to sidecar from outside mesh?

hzxuzhonghu replied:
Agree with @ramaraochavali , this appears to be confusing to me as well. We do not need to distinguish internal and external like this. I doubt: do we to allow setting different timeout/lb policy for internal and external in the future with `ExternalTrafficPolicy` and `InternalTrafficPolicy`? This seems not a good design pattern.

kfaseela replied:
@ramaraochavali the description of mesh_MTLS says   "Configuration of mTLS for traffic between workloads within the mesh.". What about traffic incoming/leaving the mesh? It did not sound matching to the description and that is why added an additional config

ramaraochavali replied:
That is what @hzxuzhonghu is eluding to - do we have to distinguish?

kfaseela replied:
@ramaraochavali : OK, I am actually fine with that. Let us hear the inputs from @howardjohn and @costinm as well. If everyone agrees, I can rework the API PR

howardjohn replied:
Is it a naming concern or a behavior concern?

mesh_mTLS: ISTIO_MUTUAL
mesh_external_TLS: not ISTIO_MUTUAL

Is it the naming that is confusing things, or do we actually want different behavior?

ramaraochavali replied:
Here is my concern

Treating non ISTIO_MUTUAL as mesh external for this use case is not correct IMO. There are many folks who use internal mesh services with MUTUAL with their own generated certs. So defining this field under Mesh_External_TLS and applying it to sidecar -> sidecar communication that uses `MUTUAL` TLS would create lot of confusion and it may even break users that set this config because now we are applying this Ecdh curves instead of Envoy defaults .

So if we want to apply to non ISTIO_MUTUAL traffic I think we need to 

1. Move this under mesh_mTLS and document that it is not applied to ISTIO_MUTUAL. So people use consciously.
2. Change the field name to additional_ecdh_curves and in the implementation add envoy default ecdh + these so that we do not break any existing MUTUAL TLS users that depend on the envoy defaults.

Not sure if #2 has any side affects.

kfaseela replied:
@ramaraochavali  (1) is okay for me. However (2) does not solve the usecase as we would need to use a different set of defaults other than what is currently present. Also the defaults are different for FIPS and non-FIPS, so it will be difficult to do (2).

ramaraochavali replied:
> However (2) does not solve the usecase as we would need to use a different set of defaults other than what is currently present. Also the defaults are different for FIPS and non-FIPS, so it will be difficult to do (2).

Ok

howardjohn replied:
@ramaraochavali for the niche users that treat MUTUAL as "mesh", why don't you just not set the field? We can easily put a comment on the field "This applies to all non ISTIO_MUTUAL TLS"?

ramaraochavali replied:
I think my main concern is, we have not documented in destination rule API or any where that if a service uses  "MUTUAL" means that is not part of mesh.  In all our apis we do not have settings that apply to MUTUAL/SIMPLE in a different category/field. So putting this configuration under mesh_external is definitely confusing to me .

 I am fine to apply it to all non ISTIO MUTUAL traffic (though I do not know the side affects of applying to ISTIO MUTUAL also) and those users will not use this field. But what is the concern in adding this field to mesh_mTLS and say this applies to all non ISTIO MUTUAL traffic (Similar to how we document specific fields in ClientTLSSettings in DR)?

howardjohn replied:
Don't we report telemetry different? and not use metadata exchange, etc?

hzxuzhonghu replied:
mostly agree. I have the same doubt, why does this setting cannot be applied to sidecar-sidecar communication?

If the target is to provide fine-grained control over traffic, I believe users even want service-grained control.

kfaseela replied:
@ramaraochavali : would it be okay, if i modify the description to remove "traffic entering or leaving the mesh", as it is not correct with the implementation. And add a description saying it is for non ISTIO MUTUAL cases alone. @howardjohn otherwise we can even move this inside mesh_mtls and then add a comment saying ecdh curves are currently applicable only for non ISTIO_MUTUAL configs, if it avoids confusions.

kfaseela replied:
@hzxuzhonghu : There is no problem with implementing it for ISTIO_MUTUAL cases, just that there is no requirement for it at the moment. We can always remove the current validation blocking this, and implement that, whenever there is a requirement?

costinm commented on mesh/v1alpha1/config.proto:
```diff
@@ -1117,13 +1117,34 @@ message MeshConfig {
+    // Optional: If specified, the TLS connection will only support the specified ECDH curves.
+    // If not specified, the default curves will be used. For additional details, refer to
+    // [Ecdh Curves](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto)
+    repeated string ecdh_curves = 2;
```
Can you clarify - what part of the TLS connection ? Is it the client cert ? Root cert ? DH exchange ?

costinm commented:
I suspected it's for the DH - my point was for the doc to be clear about
it, because curves are also used in the client cert or root cert.

And if they are used for TLS DH - most configs and APIs use one of the
standard-defined specs that include DH and cipher/MAC.

On Tue, Feb 14, 2023 at 12:42 PM Kuat ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In mesh/v1alpha1/config.proto
> <https://github.com/istio/api/pull/2543#discussion_r1106348591>:
>
> >    }
>
> +  // The below configuration parameters can be used to specify TLSConfig for in-mesh, ingress and egress TLS traffic.
> +  // For example, a user could enable min TLS version for mesh traffic and specify a curve for all egress gateways like below:
> +  // meshConfig:
> +  //   meshMTLS:
> +  //     minProtocolVersion: 1.3
> +  //   egress:
> +  //     ecdhCurves:
> +  //       - P-256
> +  //       - P-512
>
> @costinm <https://github.com/costinm> These should manifest as DH
> parameters KeyExchanges message, for the ephemeral elliptic curve DH
> exchange. Curves are separate from cipher suites, which are in the Hello
> messages I think.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/api/pull/2543#discussion_r1106348591>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAUR2U2YDZ72PVS7EUBCPDWXPU2FANCNFSM6AAAAAARWPL53Q>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

kyessenov commented:
@costinm Agree about clarity can we update the points-of-use, server tls settings and destination rules to mention that these defaults apply if unspecified?

There's no standard list of curves. FIPS limits the use of a very popular (wireguard) X25519 curve.

costinm commented:
AFAIK - there is a standard list of curves, since the handshake must
include the list with identifiers. That's in the TLS RFC AFAIK.
If it's not in a RFC and doesn't have an ID registered - it can't be used.

On Tue, Feb 14, 2023 at 2:21 PM Kuat ***@***.***> wrote:

> @costinm <https://github.com/costinm> Agree about clarity can we update
> the points-of-use, server tls settings and destination rules to mention
> that these defaults apply if unspecified?
>
> There's no standard list of curves. FIPS limits the use of a very popular
> (wireguard) X25519 curve.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/api/pull/2543#issuecomment-1430455878>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAUR2UEXGB5YP7V34SBE5DWXQAOLANCNFSM6AAAAAARWPL53Q>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

kfaseela commented:
> I suspected it's for the DH - my point was for the doc to be clear about it, because curves are also used in the client cert or root cert. And if they are used for TLS DH - most configs and APIs use one of the standard-defined specs that include DH and cipher/MAC.

Have clarified the doc to indicate that this is for DH exchange

bleggett commented:
> AFAIK - there is a standard list of curves, since the handshake must include the list with identifiers. That's in the TLS RFC AFAIK. If it's not in a RFC and doesn't have an ID registered - it can't be used.
> […](#)
> On Tue, Feb 14, 2023 at 2:21 PM Kuat ***@***.***> wrote: @costinm <https://github.com/costinm> Agree about clarity can we update the points-of-use, server tls settings and destination rules to mention that these defaults apply if unspecified? There's no standard list of curves. FIPS limits the use of a very popular (wireguard) X25519 curve. — Reply to this email directly, view it on GitHub <[#2543 (comment)](https://github.com/istio/api/pull/2543#issuecomment-1430455878)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAAUR2UEXGB5YP7V34SBE5DWXQAOLANCNFSM6AAAAAARWPL53Q> . You are receiving this because you were mentioned.Message ID: ***@***.***>

This? https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml#tls-parameters-10

As mentioned in the other PR by @kyessenov, AFAICT the constraining factor is whatever the Envoy binary we happen to be using supports, not the global IANA curve list.

Given that the sibling PR already has a basic guard to catch bad config strings and we are simply passing thru to Envoy no matter what, I think this is good enough unless/until Envoy adds the ability to self-report the curves it supports, or something.

kyessenov commented:
Feel free to file an issue to Envoy about better self-reporting or telemetry for TLS parameters. It's subtle and difficult to diagnose when it goes wrong, and can significantly impact performance numbers when choosing slow curves and large keys.

costinm commented:
On Thu, Feb 16, 2023 at 1:22 PM Ben Leggett ***@***.***>
wrote:

> AFAIK - there is a standard list of curves, since the handshake must
> include the list with identifiers. That's in the TLS RFC AFAIK. If it's not
> in a RFC and doesn't have an ID registered - it can't be used.
> … <#m_-8628898704991564857_>
> On Tue, Feb 14, 2023 at 2:21 PM Kuat *@*.*> wrote: @costinm
> <https://github.com/costinm> https://github.com/costinm
> <https://github.com/costinm> Agree about clarity can we update the
> points-of-use, server tls settings and destination rules to mention that
> these defaults apply if unspecified? There's no standard list of curves.
> FIPS limits the use of a very popular (wireguard) X25519 curve. — Reply to
> this email directly, view it on GitHub <#2543 (comment)
> <https://github.com/istio/api/pull/2543#issuecomment-1430455878>>, or
> unsubscribe
> https://github.com/notifications/unsubscribe-auth/AAAUR2UEXGB5YP7V34SBE5DWXQAOLANCNFSM6AAAAAARWPL53Q
> <https://github.com/notifications/unsubscribe-auth/AAAUR2UEXGB5YP7V34SBE5DWXQAOLANCNFSM6AAAAAARWPL53Q>
> . You are receiving this because you were mentioned.Message ID: @.*>
>
> This?
> https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml#tls-parameters-10
>
> As mentioned in the other PR by @kyessenov <https://github.com/kyessenov>,
> AFAICT the constraining factor is whatever the Envoy binary we happen to be
> using supports, not the global IANA curve list.
>
> Given that the sibling PR already has a basic guard to catch bad config
> strings and we are simply passing thru to Envoy no matter what, I think
> this is good enough unless/until Envoy adds the ability to self-report the
> curves it supports, or something.
>

1. Are you saying Envoy can't ignore ciphers/curves it doesn't support ?
This is a list of user preferences - if envoy doesn't support one, it
should be dropped
by Istiod ( we build the proxy image so we should know what we support -
with some override to allow users with custom envoy ), or ignored by envoy.
I don't
think we need a mechanism for envoy to self-report - it's already complex
enough...

2. My point really was that I would like to know what other servers (
NGinx, HAproxy, Apache, etc) use to configure TLS. I don't remember this
pattern of
configuring the TLS stack by passing the curve names. This is an API - and
it is useful to be consistent and not invent APIs without a good reason.
The TLS
settings are particularly difficult for a lot of users - and on top of
this, it is a pretty rare use case AFAIK.

3. Keep in mind we are working on the new K8S Gateway and the semantics of
DestinationRule are particularly miss-matched ( subsetting is getting
replaced, cross namespace is changed, etc) and is one of the most confusing
APIs we have. How would this feature be exposed by other
implementations of Gateway/GAMMA ?   Is the use case that a user needs to
set EC curves for individual destinations ( siteA use ec256, siteB use
 X25519 ) ? If FIPS is a requirement, would it be ok to just restrict
X25519 on a cluster/namespace/workload level - is there a case for some
workload
to not be FIPS compliant when talking with a site ?

I'm not saying we should not support this feature - we support even less
frequently used features, and FIPS is important  - but it is useful to get
a more clear
understanding of what others are using and if maybe we can support this API
without changing DestinationRule which has pretty complex composition
 and semantics. Was there a doc covering those things - I remember we
talked in the past, but in disparate PRs/features.


Message ID: ***@***.***>
>

kfaseela commented:
> On Thu, Feb 16, 2023 at 1:22 PM Ben Leggett ***@***.***> wrote:
> AFAIK - there is a standard list of curves, since the handshake must include the list with identifiers. That's in the TLS RFC AFAIK. If it's not in a RFC and doesn't have an ID registered - it can't be used. … <#m_-8628898704991564857_> On Tue, Feb 14, 2023 at 2:21 PM Kuat *@*.*> wrote: @costinm <https://github.com/costinm> https://github.com/costinm <https://github.com/costinm> Agree about clarity can we update the points-of-use, server tls settings and destination rules to mention that these defaults apply if unspecified? There's no standard list of curves. FIPS limits the use of a very popular (wireguard) X25519 curve. — Reply to this email directly, view it on GitHub <#2543 (comment) <[#2543 (comment)](https://github.com/istio/api/pull/2543#issuecomment-1430455878)>>, or unsubscribe https://github.com/notifications/unsubscribe-auth/AAAUR2UEXGB5YP7V34SBE5DWXQAOLANCNFSM6AAAAAARWPL53Q <https://github.com/notifications/unsubscribe-auth/AAAUR2UEXGB5YP7V34SBE5DWXQAOLANCNFSM6AAAAAARWPL53Q> . You are receiving this because you were mentioned.Message ID: @.*> This? https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml#tls-parameters-10 As mentioned in the other PR by @kyessenov <https://github.com/kyessenov>, AFAICT the constraining factor is whatever the Envoy binary we happen to be using supports, not the global IANA curve list. Given that the sibling PR already has a basic guard to catch bad config strings and we are simply passing thru to Envoy no matter what, I think this is good enough unless/until Envoy adds the ability to self-report the curves it supports, or something.
> 1. Are you saying Envoy can't ignore ciphers/curves it doesn't support ? This is a list of user preferences - if envoy doesn't support one, it should be dropped by Istiod ( we build the proxy image so we should know what we support - with some override to allow users with custom envoy ), or ignored by envoy. I don't think we need a mechanism for envoy to self-report - it's already complex enough... 2. My point really was that I would like to know what other servers ( NGinx, HAproxy, Apache, etc) use to configure TLS. I don't remember this pattern of configuring the TLS stack by passing the curve names. This is an API - and it is useful to be consistent and not invent APIs without a good reason. The TLS settings are particularly difficult for a lot of users - and on top of this, it is a pretty rare use case AFAIK. 3. Keep in mind we are working on the new K8S Gateway and the semantics of DestinationRule are particularly miss-matched ( subsetting is getting replaced, cross namespace is changed, etc) and is one of the most confusing APIs we have. How would this feature be exposed by other implementations of Gateway/GAMMA ? Is the use case that a user needs to set EC curves for individual destinations ( siteA use ec256, siteB use X25519 ) ? If FIPS is a requirement, would it be ok to just restrict X25519 on a cluster/namespace/workload level - is there a case for some workload to not be FIPS compliant when talking with a site ? I'm not saying we should not support this feature - we support even less frequently used features, and FIPS is important - but it is useful to get a more clear understanding of what others are using and if maybe we can support this API without changing DestinationRule which has pretty complex composition and semantics. Was there a doc covering those things - I remember we talked in the past, but in disparate PRs/features. Message ID: ***@***.***>
> […](#)

@costinm : I think we discussed this already, and the PR has been already updated to remove extensions from DR/Gateway. The current PR only has the ecdhCurves added to MeshConfig. Envoy throws ERROR if we configure an unsupported ECDHCurve, on top of that we have added warnings in istio when user puts an unrecognized curve in MeshConfig. Hope that is enough?

costinm reviewed (approved):
My mistake - not sure what happened there, clicked on an old commit which had DR. Sorry, too many context switches or old tabs.No concerns with MeshConfig - but I have an allergic reaction to DR :-), and I thought we went back for some reasons. Will be more careful.

I didn't know Envoy throws ERROR - that explains a lot, I think it is a bug. I am still confused:  
if I specify a curve in mesh config and envoy doesn't have it - will the entire mesh go down ?

In any case - it should solve your use case, and it's an alpha API so we can adjust if envoy makes some improvements.

howardjohn commented:
>if I specify a curve in mesh config and envoy doesn't have it - will the entire mesh go down ?

It will NACK so eventually - yes. We should filter in istiod as well (we already do it for cipherSuites), although it isn't foolproof since different builds of envoy have different things allowed.

istio-testing commented:
PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

kfaseela commented:
> My mistake - not sure what happened there, clicked on an old commit which had DR. Sorry, too many context switches or old tabs.No concerns with MeshConfig - but I have an allergic reaction to DR :-), and I thought we went back for some reasons. Will be more careful.
> 
> I didn't know Envoy throws ERROR - that explains a lot, I think it is a bug. I am still confused: if I specify a curve in mesh config and envoy doesn't have it - will the entire mesh go down ?
> 
> In any case - it should solve your use case, and it's an alpha API so we can adjust if envoy makes some improvements.

Thanks @costinm ! I was worried that I was still missing something :)
